### DISCUSSION TITLE
Flux bootstrap bitbucket gives panic error for a repo running on Atlassian servers

### DISCUSSION THREAD

**Discussion by compots** [fluxcd/flux2#2338] [General] [answered]
Although I am not panicking, it would seem Flux does. 

I am using Flux v0.25.3 with Bitbucket running on Atlassian's servers.
I have an existing private repository under a workspace and a project folder for flux.

When using the old way of bootstrapping everything works well where the password is an "App Password" for that relevant user and has access to the repository.
`flux bootstrap git --url=https://bitbucket.org/<workspace name>/<repo name>.git --username=<username> --password=<password> --token-auth --branch=master --path=clusters/<env name>`

When I try to do something similar with the bootstrap using the bitbucket option I get a panic:
`export BITBUCKET_TOKEN=<password>`
`flux bootstrap bitbucket-server --owner=<project name> --username=<username>  --repository=<reponame> --hostname=bitbucket.org --token-auth --branch=master --path=clusters/<env name>`

> ► connecting to bitbucket.org
> panic: runtime error: invalid memory address or nil pointer dereference
>[signal 0xc0000005 code=0x0 addr=0x0 pc=0x1415d95]
>
>goroutine 1 [running]:
>github.com/fluxcd/flux2/internal/bootstrap.(*GitProviderBootstrapper).reconcileOrgRepository(0xc0004ca0d0, {0x1cd1ab8, 0xc000177080})
>        /home/runner/work/flux2/flux2/internal/bootstrap/bootstrap_provider.go:278 +0xc15
> ....

I looked at the code but my understanding of GO is not a go-er. Perhaps orgRef.String() is throwing the nil pointer?

Anyway, I thought perhaps the owner option needs the project's key instead of its name (which for me had a space in it so I had to put it in quotes). 
I also thought that the workspace needs to be part of the repository name (i.e. --repository=<workspace name>/<repo name>) but this didn't help. I saw this because I tried to run the command using the --personal option and saw that Flux does not panic but tries to use the v1.0 api for Bitbucket to connect and reveals the url that it is trying. It doesn't have the workspace name in it.

Should I just be sticking to the old way of doing things with Bitbucket or should I persist with the bootstrapping with the bitbucket-server option?

**Comment by souleb** [comment, marked as answer]
The panic is being solved here : https://github.com/fluxcd/flux2/pull/2336. Can you try with that? Owner is the project name

**Reply by compots** [reply]
Thank you for your work in Flux. I am loving the gitops approach for K8s deployments.

However, after the upgrade I'm still getting the same panic as before for the bootstrap command given above for the bitbucket server. 
My flux version output now is:

`flux: v0.26.0`
`helm-controller: v0.15.0`
`image-automation-controller: v0.19.0`
`image-reflector-controller: v0.15.0`
`kustomize-controller: v0.19.1`
`notification-controller: v0.20.1`
`source-controller: v0.20.1`

The panic solved for #2336 seems to be for something else but was happy to go along with the thought process of the upgrade fixing things. For the record I am using extra options for my bootstrap command and am happy to be corrected on its usage:

`flux bootstrap bitbucket-server --verbose --owner=<projectname> --username=<username> --repository=flux --hostname=bitbucket.org --token-auth --branch=master --components-extra=image-reflector-controller,image-automation-controller --interval=5m0s --path=<pathname>`

My project name has a space in it so I am using double quotes around the two words but did try with %20 instead of a space and then tried project key and then back to workspace name. Alas, my attempts showed the exact same error on the same line number.

Any thoughts?

**Reply by souleb** [reply]
can you provide your bitbucket version?

**Reply by compots** [reply]
I assume it's 7.x as you have to move self hosted instances to version 7 before migration. As to the minor version number I am struggling to find a webpage that tells me.

**Reply by souleb** [reply]
please also provides the screenshot of your error.

**Reply by compots** [reply]
Thanks for looking into this. 

I noted that the original install didn't manage to create the Flux pods - they were just hanging there in container creating state - so I have done a complete flux uninstall and then tried the same bootstrap command as given above. See screenshot attached. To get my system working I went back to `flux bootstrap git ... blah blah blah` command. In this last case I have used the PAT that was put in the environment variable and it all works installs with checks passing.

As a side note, I am seeing an error in the image-automation-controller "unable to clone: User". My research on that suggests that perhaps I have network issues re [439](https://github.com/fluxcd/source-controller/issues/439) and #2307. I was seeing the same symptoms as mentioned in those issues / discussions. Everything working fine then suddenly image tags are not updated - or they take hours to get updated. I'm running K8s in Azure so I'm trying to hunt down if anything changed in Microsoft's end. I've also increased the time interval in Flux for checking Bit Bucket in the hope that I don't get throttled by Atlassian servers in case that was the cause of the problem.

![Flux-0 26 0-BitBucket org-Error](https://user-images.githubusercontent.com/13504937/152485829-88199c9d-4bd4-43da-af17-d45ef2215c7b.jpg)

**Reply by souleb** [reply]
Thanks, I see now where is the panic issue.

Another question, I see on your screenshot `bitbucket.org`. Are you using bitbucket datacenter? If so, can you confirm that this api call works?

```sh
curl $BITBUCKET_SERVER_URL/rest/api/1.0/projects/
```

**Reply by compots** [reply]
Good to hear about the panic. I can raise this as an issue if you prefer it. Just didn't want to add a lot of noise if the problem was really on my end.

I am using BitBucket in the cloud which is in Atlassian's datacentre. 

The api call does come back with a html page saying the Resource Not Found. 

I was reading [here](https://support.atlassian.com/bitbucket-cloud/docs/use-bitbucket-rest-api-version-1/) that the api v1.0 is somewhat limited for BitBucket in the cloud.

**Reply by stefanprodan** [reply]
We don't support BitBucket Cloud, only BitBucket Server aka Stash.

**Reply by souleb** [reply]
The panic is fixed :) https://github.com/fluxcd/flux2/pull/2380

Unfortunately the `bitbucket server` bootstrap only support bitbucket datacenter with the 1.0 API. Other Bitbucket services use 2.0 API, and we do not support that yet. It is in the doc here https://fluxcd.io/docs/installation/#bitbucket-server-and-data-center.

**Reply by compots** [reply]
Ah so I was right ... the OrgRef.tostring() was the culprit. 
Thanks for the fix and the link to the documation on BitBucket in the cloud. 

I assume using the `flux bootstrap git --url=https://bitbucket.org/... blah blah` command will still work even though BitBucket in the cloud will turn off the use of username / password in March 2022. Or will I be needing to migrate to ssh?

**Reply by souleb** [reply]
have you tried with a personal access token in place of the password?

**Reply by compots** [reply]
Yes, PAT is what I'm placing in the password field. I guess I'll find out what breaks in March.

**Reply by souleb** [reply]
it will continue to work, a personal access token is not a password.
